File: RungeKutta4/src/Particle.java
//TODO: format numbers 2 decimal places, calculate time, create thread to run code, create graphics to plot points

import java.awt.*;

public class Particle {
	
	private Point position;
	private Point position2;
	private Point position3;
	private Point position4;
	private Point velocity;
	private Point velocity2;
	private Point velocity3;
	private Point velocity4;
	private int mass = 1;
	private int g = -10;
	private double oldTime;;
	private static double dt = 1;
	private double energy;
	
	public Particle(int x, int y, int vx, int vy)  {
		//draw graphic
		//initialize position 	p1 = x(n)
		 position = new Point(x, y);
		 position2 = new Point();
		 position3 = new Point();
		 position4 = new Point();
		//initialize velocity  v1 = v(n)
		 velocity = new Point(vx, vy);
		 velocity2 = new Point();
		 velocity3 = new Point();
		 velocity4 = new Point();
		 
	}
	
	public double getX() {
		return position.x;
	}
	
	
	public double getY() {
		return position.y;
	}
	
	
	public double getVx() {
		return velocity.x;
	}
	
	
	public double getVy() {
		return velocity.y;
	}
	

	public double getDt() {
		return dt;
	}
	
	public double getEnergy() {
		return energy;
	}
	
	private Point getAcceleration(Point p, Point v) {
		//calculate acceleration in the x and y direction using f = ma
		//forces in y:
		return new Point(0, g);		//only accounting for the force of gravity
	}
	
	//RK4 calculations to run every frame/timestep
	public void move() {
		
		calcRK4();
		//checkBoundaries();
		//calcEnergy();
		//draw graph of trajectory and energy
		
	}
	
	public void calcRK4() {
		
		//set time = current time (getTimer())
		//set dt = current time - old time / 1000 to get milliseconds
		//set old time to new time
		//test value
		//dt = .01;
		
		//a1 = a(p1, v1)
		Point acceleration = getAcceleration(position, velocity);	
				
		//p2 = p1 + v1 * dt/2  (average t)	
				
		position2.x = (int) (position.x + velocity.x * .5 * dt);  
		position2.y = (int) (position.y + velocity.y * .5 * dt);
					
		//v2 = v1 + a1 * dt/2
				
		velocity2.x = (int) (velocity.x + acceleration.x * .5 * dt);
		velocity2.y = (int) (velocity.y + acceleration.y * .5 * dt);
					
		//a2 = a(p2, v2)
		Point acceleration2 = getAcceleration(position2, velocity2);	
					
		//p3 = p1 + v2 * dt/2
				
		position3.x = (int) (position.x + velocity2.x * .5 * dt); 
		position3.y = (int) (position.y + velocity2.y * .5 * dt);
				
		//v3 = v1 + a2 * dt/2
				
		velocity3.x = (int) (velocity.x + acceleration2.x * .5 * dt);
		velocity3.y = (int) (velocity.y + acceleration2.y * .5 * dt);
					
		//a3 = a(p3, v3)
		Point acceleration3 = getAcceleration(position3, velocity3);
				
		//p4 = p1 + v3 * dt
				
		position4.x = (int) (position.x + velocity3.x * dt);
		position4.y = (int) (position.y + velocity3.y * dt);
				
		//v4 = v1 + a3 * dt
				
		velocity4.x = (int) (velocity.x + acceleration3.x * dt);
		velocity4.y = (int) (velocity.y + acceleration3.y * dt);
					
		//a4 = a(p4, v4)
		Point acceleration4 = getAcceleration(position4, velocity4);
				
		//update current position and velocity vectors with the previously calculated values 
		//x(n + 1) = x(n) + (v1 + 2v2 + 2v3 + v4) dt / 6
		position.x += (velocity.x + 2 * velocity2.x + 2 * velocity3.x + velocity4.x) * dt / 6;
		position.y += (velocity.y + 2 * velocity2.y + 2 * velocity3.y + velocity4.y) * dt / 6;
				
		//v(n + 1) = v(n) + (a1 + 2a2 + 2a3 + a4) dt / 6
		velocity.x += (acceleration.x + 2 * acceleration2.x + 2 * acceleration3.x + acceleration4.x) * dt / 6;				
		velocity.y += (acceleration.y + 2 * acceleration2.y + 2 * acceleration3.y + acceleration4.y) * dt / 6;
		
		
	}
	
	public void checkBoundaries() {

			
	}
	
	public void calcEnergy() {
		double v = Math.sqrt(velocity.x * velocity.x + velocity.y * velocity.y);
		
		//E = K + U
		energy = Math.round((.5 * mass * Math.pow(v, 2)) + (mass * g * -1 * position.y));
		
	}
	
	
	public static void main(String[] args) {
		
		Particle p = new Particle(0, 10, 10, 0);
		
		for (double i = 0; i < 10; i+=dt) {
			//put this first to print out initial values at zero
			p.calcEnergy();
			
			System.out.println("t:" + i + "\tx:" + p.getX() + "\ty:" +p.getY() + 
					"\tvx:" + p.getVx() + "\tvy:" + p.getVy() + "\tenergy:" + p.getEnergy());
			
			p.move();
		}
		
	}
	
}

File: public/js/Particle.js
/********************************************************************
 * PARTICLE
 ********************************************************************
 * Provides the blueprint for creating objects that will be used in the world. Object 
 * is modeled after a point particle of uniform density.  Uses the 4th order Runge Kutta 
 * scheme to calculate the trajectory of the particle given an initial position and velocity.
 * Other numerical integration schemes can be implemented in the move functionality.
 * Initial velocity is input by the user which may be derived from a specified force at a
 * specified angle. Also handles collision detection. 
 * Example use: 
 * 1. instantiation: var p = new Physics.Particle(1, 2, 3, 4) 
 * 2. creating a timer: var mover = new Physics.timer(this) particle.start()
 *
 */

 var Physics = Physics || {};
 
 Physics.Body = (function () {
	var radius = 80;	
	//private static attribute initialized to -1 so that the first id will be 0 
	var nextID = 0;	
	
	//private static helper method to generate random numbers for setting x and y position
	var randomX = function () {
		var maxX = Physics.World.getWidth() - radius;
		var x = Math.floor(Math.random() * maxX);	//find a number between the length of the radius and wall
		if(x < radius) {		//if in collision with left edge
			x = radius;				//reposition to left edge
		}
		return x;
	};

	var randomY = function () {
		var maxY = Physics.World.getHeight() - radius;
		var y = Math.floor(Math.random() * maxY);
		if(y < radius) {	//if in collision with ceiling
			y = radius;
		}
		return y;
	};

	
	//constructor to create new particles
	return function () {
		//private attributes 
		var x, y, vx, vy;
		var mass = 1;		//default to 1 for easy computation.  Should be initialized in constructor.
		var radius = 80;	
		var energy;			//kinetic energy at any time interval
		var skipUpdate;		//used if the particle is stationary 
		var skipDraw;
		var id;
		var svgns = "http://www.w3.org/2000/svg";	//svg namespace for initializing circles
		var circleID;
			
		this.getX = function () {
			return x;
		};
		
		this.setX = function (xi) {
			x = xi;	//update variable for calculations
		};
		
		this.getY = function () {
			return y;
		};
		
		this.setY = function (yi) {
			y = yi;
		};
		
		this.getVx = function () {
			return vx;
		};
		
		this.setVx = function(vxi) {
			vx = vxi;
		};
		
		this.getVy = function () {
			return vy;
		};
		
		this.setVy = function (vyi) {
			vy = vyi;
		};
		
		this.getRadius = function () {
			return radius;
		};
		
		this.setRadius = function (r) {
			radius = r;
		};
		
		this.getEnergy = function () {
			return Math.round(energy);
		};

		this.setEnergy = function (e) {
			energy = e;
		}
		
		this.getID = function () {
			return id;
		}
		
		this.setID = function (idNum) {
			id = idNum;
		};
		
		this.getCircleID = function () {
			return circleID;
		};
		
		//calculates total mechanical energy at any given time interval using 1/2 m v^2 + mgh 
		//where m = mass, v = velocity, g = gravity, and h = height or the y value.
		this.calcEnergy = function () {
			//find the magnitude of the velocity 
			var v = Math.sqrt(vx * vx + vy * vy);
			
			//Energy = Kinetic Energy + Potential Energy
			var e = (.5 * Math.pow(v, 2)) + (Physics.World.getG() * -1 * y); //multiply by -1 because gravity and height are opposites
			this.setEnergy(e);
		};
		
		//deallocates particle from memory, removes element from world
		this.destroy = function () {
			var circle = document.getElementById(this.circleID);
			document.getElementById('world').removeChild(circle);
		};
		
		//takes a particle object and updates position and velocity using specified scheme, 
		this.move = function () {
			Physics.RK4.calcRK4(this);
			//this.update();
			//console.log("I moved");
		};
		
		//initializes particle properties and draws it
		this.init = function () {
			this.setX(randomX());
			this.setY(randomY());
			this.setVx(Math.random() * 1000);
			this.setVy(Math.random() * 1000);
			this.setID(nextID);
			this.setEnergy(0);
			nextID++;
			this.drawSVG();
			console.log("particle initialized");
		};
		
		this.drawSVG = function () {
			//set the circle's attributes to the particle's attributes
			var circle = document.createElementNS(svgns, 'circle');
			circle.setAttribute('cx', x);
			circle.setAttribute('cy', y);
			circle.setAttribute('r', radius);
			circle.setAttribute('stroke', 'black');
			circle.setAttribute('stroke-width', 2);
			circle.style.fill = "white";
			//add an id to the circle so that it can be called by id
			circleID = "circle" + id;
			circle.setAttribute('id', circleID);
			document.getElementById("world").appendChild(circle);
		}
		//visualizes particle as a circle on the screen
		this.draw = function () {
			//set the circle's attributes to the particle's attributes
			var circle = document.createElement('DIV');
			circle.style.left = this.x;
			circle.style.top = this.y;
			circleID = "circle" + id;
			circle.setAttribute('id', circleID);
			circle.setAttribute('class', "circle");
			//add the circle to the svg canvas
			document.getElementById("world").appendChild(circle);
		};
		
		// draws object, updates momentum and energy on every keyframe or when 
		//setting object prior to animating, or when refreshing.  
		this.update = function () {
			
			document.getElementById(circleID).setAttribute('cx', x);
			document.getElementById(circleID).setAttribute('cy', y);
			this.calcEnergy();
			//stopping condition
			//when energy is below n stop the animation.
			if(this.getEnergy() < 50) {
				this.setVx(0);
				this.setVy(0);
			}
			//console.log(this.getEnergy());
			
		};
		
		//calls particle’s destroy method. Term “cleanup” used for convenience and consistency
		//in order to iterate over the cleanup method of all objects and managers
		this.cleanup = function () {
			this.destroy();
		};
		
		
	};//end function
	
	
	
})();	
/*
var p = new Physics.Body(); 
p.init();
p.setX(80);
p.setY(80);
console.log("x:" + p.getX() + " y:" + p.getY() + " vx:" + p.getVx() + " vy:" + p.getVy());
//var p = new Physics.Body(); 
//console.log("x:" + p.getX() + " y:" + p.getY() + " vx:" + p.getVx() + " vy:" + p.getVy());

//p.init();
//p.init();
//p.init();
//p.init();
*/
